domain/schema/model/sql/0007-charm.sql

```diff
@@ -1,27 +1,398 @@
+    assumes_blob        TEXT,
+    CONSTRAINT          fk_charm_run_as_kind_charm
+        FOREIGN KEY     (run_as_id)
+        REFERENCES      charm_run_as_kind(id)
+);
+
+CREATE UNIQUE INDEX idx_charm_name
+    ON charm (name);
+
+-- The charm_state table exists to store the availability of a charm. The
+-- fact that the charm is in the database indicates that it's a placeholder.
+-- Updating the available flag to true indicates that the charm is now
+-- available for deployment.
+-- This is exists as a separate table as the charm table models the charm
+-- metadata and the goal state of the charm. The charm_state table models the
+-- internal state of the charm.
+CREATE TABLE charm_state (
+    charm_uuid          TEXT NOT NULL,
+    -- Available is a flag that indicates whether the charm is available for
+    -- deployment.
+    available           BOOLEAN,
+    CONSTRAINT          fk_charm_state_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid)
+);
+
+CREATE TABLE charm_source (
+    id       INT PRIMARY KEY,
+    name     TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_charm_source_name
+    ON charm_source (name);
+
+INSERT INTO charm_source VALUES
+    (0, 'local'),
+    (1, 'ch');
+
+CREATE TABLE charm_origin (
+    charm_uuid          TEXT NOT NULL,
+    source_id           INT,
+    id                  string,
+    hash                string,
+    revision            INT,
+    CONSTRAINT          fk_charm_source_source
+        FOREIGN KEY     (source_id)
+        REFERENCES      charm_source(id),
+    CONSTRAINT          fk_charm_origin_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid)
+);
+
+CREATE TABLE charm_channel (
+    charm_uuid          TEXT NOT NULL,
+    track               TEXT,
+    risk                TEXT,
+    branch              TEXT,
+    CONSTRAINT          fk_charm_channel_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid)
+);
+
+CREATE TABLE architecture (
+    id       INT PRIMARY KEY,
+    name     TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_architecture_name
+    ON architecture (name);
+
+INSERT INTO architecture VALUES
+    (0, 'amd64'),
+    (1, 'arm64'),
+    (2, 'ppc64el'),
+    (3, 's390x'),
+    (4, 'riscv64');
+
+CREATE TABLE charm_platform (
+    charm_uuid          TEXT NOT NULL,
+    os                  TEXT,
+    channel             TEXT,
+    architecture_id     TEXT,
+    CONSTRAINT          fk_charm_platform_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    CONSTRAINT          fk_charm_platform_architecture
+        FOREIGN KEY     (architecture_id)
+        REFERENCES      architecture(id)
+);
+
+CREATE TABLE hash_kind (
+    id       INT PRIMARY KEY,
+    name     TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_hash_kind_name
+    ON hash_kind (name);
+
+INSERT INTO hash_kind VALUES
+    (0, 'sha256');
+
+CREATE TABLE charm_hash (
```

Comment:
I wonder whether we should fold this into the charm_state table (don't necessarily retain that table name). What I mean is that the hash is known when the download is done and the charm is available. So there's a logical grouping of concerns there.